plugins/woocommerce-blocks/assets/js/blocks/checkout/edit.tsx

```diff
@@ -69,13 +65,93 @@ export const Edit = ( {
 		showFormStepNumbers = false,
 	} = attributes;
 
+	const defaultFields = useSelect( ( select ) => {
+		const settings = select(
+			coreStore as unknown as string
+		).getEditedEntityRecord( 'root', 'site' ) as Record< string, string >;
+
+		const fieldsWithDefaults = {
+			phone: 'optional',
+			company: 'hidden',
+			address_2: 'optional',
+		} as const;
+
+		return {
+			...defaultFieldsSetting,
+			...Object.fromEntries(
+				Object.entries( fieldsWithDefaults ).map(
+					( [ field, defaultValue ] ) => {
+						const value =
+							settings[
+								`woocommerce_checkout_${ field }_field`
+							] || defaultValue;
+						return [
+							field,
+							{
+								...defaultFieldsSetting[ field ],
+								required: value === 'required',
+								hidden: value === 'hidden',
+							},
+						];
+					}
+				)
+			),
+		};
+	} );
+
+	// These state objects are in place so the required toggles remember their last value after being hidden and shown.
+	const [ lastPhoneRequired, setLastPhoneRequired ] = useState(
+		defaultFields.phone.required
+	);
+	const [ lastCompanyRequired, setLastCompanyRequired ] = useState(
+		defaultFields.company.required
+	);
+	const [ lastAddress2Required, setLastAddress2Required ] = useState(
+		defaultFields.address_2.required
+	);
+
+	useEffect( () => {
+		if ( ! defaultFields.phone.hidden ) {
+			setLastPhoneRequired( defaultFields.phone.required );
+		}
+		if ( ! defaultFields.company.hidden ) {
+			setLastCompanyRequired( defaultFields.company.required );
+		}
+		if ( ! defaultFields.address_2.hidden ) {
+			setLastAddress2Required( defaultFields.address_2.required );
+		}
+	}, [
+		defaultFields.phone.hidden,
+		defaultFields.company.hidden,
+		defaultFields.address_2.hidden,
+		defaultFields.phone.required,
+		defaultFields.company.required,
+		defaultFields.address_2.required,
+	] );
```

Comment:
Doesn't seem scalable as we want to add those control to more fields later on but we solve it once we reach it

Reply:
Yeah I had to do this because of the way the UI works; without tracking state you would lose the previous checked state when toggling visibility. Thats because the visibility and required status is tracked within a single entitiy.